File: DynamicKey/AgoraDynamicKey/dart/lib/src/stt_token_builder.dart
import 'access_token.dart';
import 'rtc_token_builder.dart';

/// Builds Token007 tokens for RTC, RTM, and STT services.
class SttTokenBuilder {
  /// Builds a Token007 that carries RTC, RTM, and STT services.
  ///
  /// [appId] is the App ID issued by Agora.
  /// [appCertificate] is the certificate of the application registered in Agora Dashboard.
  /// [channelName] is the unique channel name for the RTC session.
  /// [rtcAccount] is the RTC user account and must not exceed 255 bytes.
  /// [rtcRole] is [RtcRole.publisher] for a broadcaster or [RtcRole.subscriber] for an audience member.
  /// [rtcTokenExpireSeconds] is the number of seconds from now until the token expires. This is the whole
  /// token expiration. The whole token is invalid after this time, even if a privilege expiration time is
  /// later.
  /// [joinChannelPrivilegeExpire] is the number of seconds from now until the join privilege expires. This
  /// value must not exceed the token expiration value; otherwise, the privilege is limited by the token
  /// expiration time.
  /// [publishAudioPrivilegeExpire] is the number of seconds from now until the audio publishing privilege
  /// expires. This value must not exceed the token expiration value; otherwise, the privilege is limited by
  /// the token expiration time.
  /// [publishVideoPrivilegeExpire] is the number of seconds from now until the video publishing privilege
  /// expires. This value must not exceed the token expiration value; otherwise, the privilege is limited by
  /// the token expiration time.
  /// [publishDataPrivilegeExpire] is the number of seconds from now until the data publishing privilege
  /// expires. This value must not exceed the token expiration value; otherwise, the privilege is limited by
  /// the token expiration time.
  /// [rtmUserId] is the RTM user account and must not exceed 255 bytes.
  /// [rtmTokenExpireSeconds] is the number of seconds from now until the RTM privilege expires. This value
  /// must not exceed the RTC token expiration value; otherwise, the RTM login privilege is limited by the
  /// token expiration time.
  ///
  /// Returns the generated RTC, RTM, and STT token.
  static String buildToken({
    required String appId,
    required String appCertificate,
    required String channelName,
    required String rtcAccount,
    required RtcRole rtcRole,
    required int rtcTokenExpireSeconds,
    required int joinChannelPrivilegeExpire,
    required int publishAudioPrivilegeExpire,
    required int publishVideoPrivilegeExpire,
    required int publishDataPrivilegeExpire,
    required String rtmUserId,
    required int rtmTokenExpireSeconds,
  }) {
    final token = AccessToken.create(
      appId,
      appCertificate,
      expire: rtcTokenExpireSeconds,
    );

    final serviceRtc = ServiceRtc(channelName, rtcAccount)
      ..addPrivilege(
        ServiceRtc.privilegeJoinChannel,
        joinChannelPrivilegeExpire,
      );
    if (rtcRole == RtcRole.publisher) {
      serviceRtc
        ..addPrivilege(
          ServiceRtc.privilegePublishAudioStream,
          publishAudioPrivilegeExpire,
        )
        ..addPrivilege(
          ServiceRtc.privilegePublishVideoStream,
          publishVideoPrivilegeExpire,
        )
        ..addPrivilege(
          ServiceRtc.privilegePublishDataStream,
          publishDataPrivilegeExpire,
        );
    }
    token.addService(serviceRtc);

    final serviceRtm = ServiceRtm(rtmUserId)
      ..addPrivilege(ServiceRtm.privilegeLogin, rtmTokenExpireSeconds);
    token.addService(serviceRtm);

    token.addService(ServiceStt());

    return token.build();
  }
}

File: DynamicKey/AgoraDynamicKey/nodejs/src/SttTokenBuilder.js
const AccessToken = require('../src/AccessToken2').AccessToken2
const ServiceRtc = require('../src/AccessToken2').ServiceRtc
const ServiceRtm = require('../src/AccessToken2').ServiceRtm
const ServiceStt = require('../src/AccessToken2').ServiceStt
const Role = require('../src/RtcTokenBuilder2').Role

class SttTokenBuilder {
    /**
     * Builds a Token007 that carries RTC, RTM, and STT services.
     * @param {*} appId The App ID issued to you by Agora.
     * @param {*} appCertificate Certificate of the application that you registered in the Agora Dashboard.
     * @param {*} channelName The unique channel name for the AgoraRTC session in the string format.
     * @param {*} rtcAccount The RTC user's account, max length is 255 bytes.
     * @param {*} rtcRole Role.PUBLISHER for a broadcaster or Role.SUBSCRIBER for an audience member.
     * @param {*} rtcTokenExpire Represented by the number of seconds elapsed since now. This is the whole
     * token expiration. The whole token is invalid after this time, even if a privilege expiration time is
     * later.
     * @param {*} joinChannelPrivilegeExpire Represented by the number of seconds elapsed since now. This
     * value must not exceed rtcTokenExpire; otherwise, the privilege is limited by the token expiration time.
     * @param {*} pubAudioPrivilegeExpire Represented by the number of seconds elapsed since now. This value
     * must not exceed rtcTokenExpire; otherwise, the privilege is limited by the token expiration time.
     * @param {*} pubVideoPrivilegeExpire Represented by the number of seconds elapsed since now. This value
     * must not exceed rtcTokenExpire; otherwise, the privilege is limited by the token expiration time.
     * @param {*} pubDataStreamPrivilegeExpire Represented by the number of seconds elapsed since now. This
     * value must not exceed rtcTokenExpire; otherwise, the privilege is limited by the token expiration time.
     * @param {*} rtmUserId The RTM user's account, max length is 255 bytes.
     * @param {*} rtmTokenExpire Represented by the number of seconds elapsed since now. This value must not
     * exceed rtcTokenExpire; otherwise, the RTM login privilege is limited by the token expiration time.
     * @return The RTC, RTM, and STT token.
     */
    static buildToken(
        appId,
        appCertificate,
        channelName,
        rtcAccount,
        rtcRole,
        rtcTokenExpire,
        joinChannelPrivilegeExpire,
        pubAudioPrivilegeExpire,
        pubVideoPrivilegeExpire,
        pubDataStreamPrivilegeExpire,
        rtmUserId,
        rtmTokenExpire
    ) {
        const token = new AccessToken(appId, appCertificate, 0, rtcTokenExpire)

        const serviceRtc = new ServiceRtc(channelName, rtcAccount)
        serviceRtc.add_privilege(ServiceRtc.kPrivilegeJoinChannel, joinChannelPrivilegeExpire)
        if (rtcRole === Role.PUBLISHER) {
            serviceRtc.add_privilege(ServiceRtc.kPrivilegePublishAudioStream, pubAudioPrivilegeExpire)
            serviceRtc.add_privilege(ServiceRtc.kPrivilegePublishVideoStream, pubVideoPrivilegeExpire)
            serviceRtc.add_privilege(ServiceRtc.kPrivilegePublishDataStream, pubDataStreamPrivilegeExpire)
        }
        token.add_service(serviceRtc)

        const serviceRtm = new ServiceRtm(rtmUserId)
        serviceRtm.add_privilege(ServiceRtm.kPrivilegeLogin, rtmTokenExpire)
        token.add_service(serviceRtm)

        token.add_service(new ServiceStt())

        return token.build()
    }
}

module.exports = {
    SttTokenBuilder
}
